pydantic/utils.py

```diff
@@ -136,6 +139,51 @@ def almost_equal_floats(value_1: float, value_2: float, *, delta: float = 1e-8)
     return abs(value_1 - value_2) <= delta
 
 
+def generate_model_signature(
+    init: Callable[..., None], fields: Dict[str, 'ModelField'], config: Type['BaseConfig']
+) -> 'Signature':
+    """
+    Generate signature for model based on its fields
+    """
+    from inspect import Parameter, Signature, signature
+
+    present_params = signature(init).parameters.values()
+    merged_params: Dict[str, Parameter] = {}
+    var_kw = None
+    use_var_kw = False
+
+    for param in islice(present_params, 1, None):  # skip self arg
+        if param.kind is param.VAR_KEYWORD:
+            var_kw = param
+            continue
+        merged_params[param.name] = param
+
+    if var_kw:  # if custom init has no var_kw, fields which are not declared in it cannot be passed through
+        allow_names = config.allow_population_by_field_name
+        for field_name, field in fields.items():
+            param_name = field.alias
+            if field_name in merged_params or param_name in merged_params:
+                continue
+            elif not param_name.isidentifier():
+                if allow_names and field_name.isidentifier():
+                    param_name = field_name
+                else:
+                    use_var_kw = True
+                    continue
+
+            # TODO: replace annotation with actual expected types once #1055 solved
+            kwargs = {'default': field.default} if not field.required else {}
+            merged_params[param_name] = Parameter(param_name, Parameter.KEYWORD_ONLY, annotation=field.type_, **kwargs)
+
+    if config.extra is config.extra.allow:
```

Comment:
As proposed in https://github.com/samuelcolvin/pydantic/pull/1034#issuecomment-592181568
```suggestion
    if config.extra is config.extra.allow:
        if not var_kw:
            raise TypeError("Using `Config.extra == Extra.allow` without **data parameter in __init__ is ambiguous")

```
@samuelcolvin, what your thoughts on this?

Reply:
I disagree with this, people should be allowed to create an init method that doesn't allow extra:
* we don't do any other checks that `__init__` matches model
* I can imagine a scenario where a super method has `extra = allow` but a submodel wants a custom init method without having to bother with modifying the config